tokio-trace/src/value.rs

```diff
@@ -0,0 +1,363 @@
+
+/// An owned value of an erased type.
+///
+/// Like `Any`, references to `OwnedValue` may attempt to downcast the value to
+/// a concrete type. However, unlike `Any`, `OwnedValue`s are constructed from
+/// types known to implement `AsValue`, providing the `fmt_value` method.
+/// This means that arbitrary `OwnedValue`s may be formatted using the erased
+/// type's `fmt_value` implementation, _even when the erased type is no longer
+/// known_.
+pub struct OwnedValue(Box<dyn AsValue>);
+
+/// Converts a reference to a `T` into a `BorrowedValue`.
+pub fn borrowed<'a>(t: &'a dyn AsValue) -> BorrowedValue<'a> {
+    BorrowedValue(t)
+}
+
+/// Wraps a type implementing `fmt::Display` so that its `Display`
+/// implementation will be used when formatting it as a `Value`.
+///
+/// # Examples
+/// ```
+/// # extern crate tokio_trace;
+/// use tokio_trace::value;
+/// # use std::fmt;
+/// # fn main() {
+///
+/// #[derive(Clone, Debug)]
+/// struct Foo;
+///
+/// impl fmt::Display for Foo {
+///     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+///         f.pad("Hello, I'm Foo")
+///     }
+/// }
+///
+/// let foo = Foo;
+/// assert_eq!("Foo".to_owned(), format!("{:?}", foo));
+///
+/// let display_foo = value::display(foo.clone());
+/// assert_eq!(
+///     format!("{}", foo),
+///     format!("{:?}", value::borrowed(&display_foo)),
+/// );
+/// # }
+/// ```
+///
+/// ```
+/// # extern crate tokio_trace;
+/// # use std::fmt;
+/// # fn main() {
+/// #
+/// # #[derive(Clone, Debug)]
+/// # struct Foo;
+/// #
+/// # impl fmt::Display for Foo {
+/// #   fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+/// #       f.pad("Hello, I'm Foo")
+/// #   }
+/// # }
+/// use tokio_trace::value::{self, Value, IntoValue};
+/// let foo = value::display(Foo);
+///
+/// let owned_value = foo.into_value();
+/// assert_eq!("Hello, I'm Foo".to_owned(), format!("{:?}", owned_value));
+///
+/// assert!(owned_value.downcast_ref::<Foo>().is_some());
+/// # }
+/// ```
+pub fn display<T>(t: T) -> DisplayValue<T>
+where
+    T: AsValue + fmt::Display,
+{
+    DisplayValue(t)
+}
+
+// ===== impl AsValue =====
+
+// Copied from `std::any::Any`.
+impl AsValue + 'static {
+    /// Returns true if the boxed type is the same as `T`
+    #[inline]
+    fn is<T: AsValue + 'static>(&self) -> bool
+    where
+        Self: 'static
+    {
+        // Get TypeId of the type this function is instantiated with
+        let t = TypeId::of::<T>();
+
+        // Get TypeId of the type in the trait object
+        let boxed = self.type_id();
+
+        // Compare both TypeIds on equality
+        t == boxed
+    }
+
+    /// Returns some reference to the boxed value if it is of type `T`, or
+    /// `None` if it isn't.
+    fn downcast_ref<T: AsValue + 'static>(&self) -> Option<&T>
+    where
+        Self: 'static,
+    {
+        if self.is::<T>() {
+            unsafe {
+                Some(&*(self as *const AsValue as *const T))
```

Comment:
this is safe because:
- we know `self` has the same type ID as the type we're casting to, so the cast should be valid.
- argument by authority: `std` does it and we endorse the same lifetime invariants that `std` does